drivers/pcie/host/pcie.c

```diff
@@ -370,6 +370,93 @@ pcie_bdf_t pcie_bdf_lookup(pcie_id_t id)
 	return PCIE_BDF_NONE;
 }
 
+/* Forward declaration needed since scanning a device may reveal a bridge */
+static bool scan_bus(int bus, const struct pcie_scan_opt *opt);
+
+static bool scan_dev(int bus, int dev, const struct pcie_scan_opt *opt)
+{
+	for (int func = 0; func <= PCIE_MAX_FUNC; func++) {
+		pcie_bdf_t bdf = PCIE_BDF(bus, dev, func);
+		uint32_t secondary = 0;
+		uint32_t id, type;
+		bool do_cb;
+
+		id = pcie_conf_read(bdf, PCIE_CONF_ID);
+		if (!PCIE_ID_IS_VALID(id)) {
+			continue;
+		}
+
+		type = pcie_conf_read(bdf, PCIE_CONF_TYPE);
+		switch (PCIE_CONF_TYPE_GET(type)) {
+		case PCIE_CONF_TYPE_STANDARD:
+			do_cb = true;
+			break;
+		case PCIE_CONF_TYPE_PCI_BRIDGE:
+			if (opt->recursive) {
+				uint32_t num = pcie_conf_read(bdf,
+							      PCIE_BUS_NUMBER);
+				secondary = PCIE_BUS_SECONDARY_NUMBER(num);
+			}
+			__fallthrough;
+		default:
+			do_cb = opt->cb_all;
+			break;
+		}
+
+		if (do_cb && !opt->cb(bdf, id, opt->cb_data)) {
```

Comment:
need to test opt->cb validity prior to calling it no?

Reply:
I can't really think of any valid use cases for using `pcie_scan()` without a callback, so I think I'd just return `-EINVAL` straight away instead of checking for NULL here.

Reply:
I've added a CHECKIF() for this